src/libopensc/card-epass2003.c

```diff
@@ -502,37 +605,78 @@ gen_init_key(struct sc_card *card, unsigned char *key_enc, unsigned char *key_ma
 	LOG_TEST_RET(card->ctx, r, "gen_init_key failed");
 
 	/* Step 1 - Generate Derivation data */
-	memcpy(data, &result[16], 4);
-	memcpy(&data[4], g_random, 4);
-	memcpy(&data[8], &result[12], 4);
-	memcpy(&data[12], &g_random[4], 4);
+	if(isFips){
+        	memset(data,0x00,15);
+        	memset(&data[11], 0x04, 1);
+        	memset(&data[14], 0x80, 1);
+        	memset(&data[15], 0x01, 1);
+        	memcpy(&data[16], g_random, 8);
+        	memcpy(&data[24], &result[12+1], 8);
+    	}
+    	else{
+		memcpy(data, &result[16], 4);
+		memcpy(&data[4], g_random, 4);
+		memcpy(&data[8], &result[12], 4);
+		memcpy(&data[12], &g_random[4], 4);
+	}
 
 	/* Step 2,3 - Create S-ENC/S-MAC Session Key */
 	if (KEY_TYPE_AES == key_type) {
-		aes128_encrypt_ecb(key_enc, 16, data, 16, exdata->sk_enc);
-		aes128_encrypt_ecb(key_mac, 16, data, 16, exdata->sk_mac);
+		if(isFips){
+        	    	r = aes128_encrypt_cmac(key_enc, 128, data, 32, exdata->sk_enc);
+            		LOG_TEST_RET(card->ctx, r, "aes128_encrypt_cmac enc failed");
+            		memset(&data[11], 0x06, 1);
+            		r = aes128_encrypt_cmac(key_mac, 128, data, 32, exdata->sk_mac);
+            		LOG_TEST_RET(card->ctx, r, "aes128_encrypt_cmac mac  failed");
+        	}else{
+			r = aes128_encrypt_ecb(key_enc, 16, data, 16, exdata->sk_enc);
+			LOG_TEST_RET(card->ctx, r, "aes128_encrypt_ecb enc  failed");
+			r = aes128_encrypt_ecb(key_mac, 16, data, 16, exdata->sk_mac);
+			LOG_TEST_RET(card->ctx, r, "aes128_encrypt_ecb mac  failed");
+		}
 	}
 	else {
-		des3_encrypt_ecb(key_enc, 16, data, 16, exdata->sk_enc);
-		des3_encrypt_ecb(key_mac, 16, data, 16, exdata->sk_mac);
+		r = des3_encrypt_ecb(key_enc, 16, data, 16, exdata->sk_enc);
+		LOG_TEST_RET(card->ctx, r, "des3_encrypt_ecb failed");
+		r = des3_encrypt_ecb(key_mac, 16, data, 16, exdata->sk_mac);
+		LOG_TEST_RET(card->ctx, r, "des3_encrypt_ecb failed");
 	}
 
-	memcpy(data, g_random, 8);
-	memcpy(&data[8], &result[12], 8);
-	data[16] = 0x80;
-	blocksize = (key_type == KEY_TYPE_AES ? 16 : 8);
-	memset(&data[17], 0x00, blocksize - 1);
+	if(isFips){
+		memset(&data[11], 0x00, 1);
+		memset(&data[14], 0x40, 1);
```

Comment:
this is again assigning one value using memset, which sounds to me like overkill. Why would normal assignment not work here?